pkg/promtail/targets/gcplog/target.go

```diff
@@ -0,0 +1,165 @@
+		Namespace: "promtail",
+		Name:      "gcplog_target_entries_total",
+		Help:      "Help number of successful entries sent to the gcplog target",
+	}, []string{"project"})
+
+	gcplogErrors = promauto.NewCounterVec(prometheus.CounterOpts{
+		Namespace: "promtail",
+		Name:      "gcplog_target_parsing_errors_total",
+		Help:      "Total number of parsing errors while receiving gcplog messages",
+	}, []string{"project"})
+)
+
+// GcplogTarget represents the target specific to GCP project.
+// It collects logs from GCP and push it to Loki.
+type GcplogTarget struct {
+	logger        log.Logger
+	handler       api.EntryHandler
+	config        *scrapeconfig.GcplogTargetConfig
+	relabelConfig []*relabel.Config
+	jobName       string
+
+	// lifecycle management
+	ctx    context.Context
+	cancel context.CancelFunc
+
+	// pubsub
+	ps   *pubsub.Client
+	msgs chan *pubsub.Message
+}
+
+// NewGcplogTarget returns the new instannce of GcplogTarget for
+// the given `project-id`. It scraps logs from the GCP project
+// and push it Loki via given `api.EntryHandler.`
+// It starts the `run` loop to consume log entries that can be
+// stopped via `target.Stop()`
+func NewGcplogTarget(
+	logger log.Logger,
+	handler api.EntryHandler,
+	relabel []*relabel.Config,
+	jobName string,
+	config *scrapeconfig.GcplogTargetConfig,
+) (*GcplogTarget, error) {
+
+	ctx, cancel := context.WithCancel(context.Background())
+
+	ps, err := pubsub.NewClient(ctx, config.ProjectID)
+	if err != nil {
+		return nil, err
+	}
+
+	target, err := newGcplogTarget(logger, handler, relabel, jobName, config, ps, ctx, cancel)
+	if err != nil {
+		return nil, err
+	}
+
+	go target.run()
+
+	return target, nil
+}
+
+func newGcplogTarget(
+	logger log.Logger,
+	handler api.EntryHandler,
+	relabel []*relabel.Config,
+	jobName string,
+	config *scrapeconfig.GcplogTargetConfig,
+	pubsubClient *pubsub.Client,
+	ctx context.Context,
+	cancel func(),
+) (*GcplogTarget, error) {
+
+	pt := &GcplogTarget{
+		logger:        logger,
+		handler:       handler,
+		relabelConfig: relabel,
+		config:        config,
+		jobName:       jobName,
+		ctx:           ctx,
+		cancel:        cancel,
+		ps:            pubsubClient,
+		msgs:          make(chan *pubsub.Message),
+	}
+
+	return pt, nil
+}
+
+func (t *GcplogTarget) run() error {
+	send := t.handler.Chan()
+
+	sub := t.ps.SubscriptionInProject(t.config.Subscription, t.config.ProjectID)
+	go func() {
+		err := sub.Receive(t.ctx, func(ctx context.Context, m *pubsub.Message) {
+			t.msgs <- m
+		})
+		if err != nil {
+			// TODO(kavi): Add proper error propagation maybe?
+			level.Error(t.logger).Log("error", err)
+			gcplogErrors.WithLabelValues(t.config.ProjectID).Inc()
+		}
+	}()
+
+	for {
+		select {
+		case <-t.ctx.Done():
+			return t.ctx.Err()
+		case m := <-t.msgs:
+			entry, err := format(m, t.config.Labels, t.config.UseIncomingTimestamp)
+			if err != nil {
+				level.Error(t.logger).Log("event", "error formating log entry", "cause", err)
+				break
```

Comment:
do you mean continue ?

Reply:
I mean only `break`. This is just break from `select`. Still we inside the `for {}`

Reply:
ho I see but shouldn't e ack it ?

Reply:
I think we shouldn't, if we ack, then we may loose the message forever. But also if we don't then we may receive same message again and again and our formatter will keep failing. 

At least in my opinion, this should be some kind of alerts for manual intervention!

I don't know whats the right option. wdyt?

Reply:
log and ack would be my choice.